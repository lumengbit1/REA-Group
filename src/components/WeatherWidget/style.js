import styled from 'styled-components';

export const Table = styled.table`
  width: 100%;

  th, td {
    text-align: center;
    padding: 10px;
    border: 1px solid black;
  }
`;
