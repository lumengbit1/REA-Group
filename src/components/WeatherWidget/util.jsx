import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Map } from 'immutable';

const withComponentLoading = (WrappedComponent) => {
  function WithWrap(props) {
    const { loading, weatherRecords } = props;

    if (loading) return 'Loading...';

    if (!weatherRecords || weatherRecords.isEmpty()) return null;

    return (
      <WrappedComponent {...props} />
    );
  }

  WithWrap.propTypes = {
    loading: PropTypes.bool,
    weatherRecords: ImmutablePropTypes.map,
  };

  WithWrap.defaultProps = {
    loading: undefined,
    weatherRecords: Map(),
  };

  return WithWrap;
};

export default withComponentLoading;
