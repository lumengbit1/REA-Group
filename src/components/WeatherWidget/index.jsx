import 'moment-timezone';
import React from 'react';
import moment from 'moment';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Map } from 'immutable';
import withComponentLoading from './util';
import { Table } from './style';

const WeatherWidget = (props) => {
  const { weatherRecords } = props;

  return (
    <Table>
      <thead>
        <tr>
          <th>Date</th>
          <th>Maximum Temperature</th>
          <th>Minimum Temperature</th>
          <th>Weather Conditions</th>
        </tr>
      </thead>

      <tbody>
        {weatherRecords.get('daily').map((weatherValue, index) => (
          <tr key={weatherValue.get('dt')}>
            <td>
              {moment().add(index, 'days').tz(weatherRecords.get('timezone')).format('dddd, Do MMM YYYY')}
            </td>
            <td>
              {`${weatherValue.get('temp').get('max')}°`}
            </td>
            <td>
              {`${weatherValue.get('temp').get('min')}°`}
            </td>
            <td>
              {weatherValue.get('weather') && weatherValue.get('weather').map((val) => val.get('description'))}
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

WeatherWidget.propTypes = {
  weatherRecords: ImmutablePropTypes.map,
};

WeatherWidget.defaultProps = {
  weatherRecords: Map(),
};

export default withComponentLoading(WeatherWidget);
