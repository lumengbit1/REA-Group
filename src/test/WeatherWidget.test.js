import React from 'react';
import 'jest-styled-components';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import WeatherWidget from '../components/WeatherWidget';
import { fromJS } from 'immutable';

Enzyme.configure({ adapter: new Adapter() });

describe('render snapshop testing', () => {
  it('1.renders correctly', () => {
    const wrapper = shallow(<WeatherWidget />);

    expect(wrapper).toMatchSnapshot();
  });

  it('2.weatherRecords props is undefined', () => {
    const wrapper = shallow(<WeatherWidget weatherRecords={undefined} />);

    expect(wrapper.html()).toBe(null);
  });

  it('3.loading props is true', () => {
    const wrapper = shallow(<WeatherWidget loading={true} />);

    expect(wrapper.text()).toBe('Loading...');
  });

  it('4.weatherRecords props has data', () => {
    const weatherData = fromJS({
      "lat": 33.44,
      "lon": -94.04,
      "timezone": "America/Chicago",
      "timezone_offset": -21600,
      "daily": [
        {
          "dt": 1618308000,
          "sunrise": 1618282134,
          "sunset": 1618333901,
          "moonrise": 1618284960,
          "moonset": 1618339740,
          "moon_phase": 0.04,
          "temp": {
            "day": 279.79,
            "min": 275.09,
            "max": 284.07,
            "night": 275.09,
            "eve": 279.21,
            "morn": 278.49
          },
          "feels_like": {
            "day": 277.59,
            "night": 276.27,
            "eve": 276.49,
            "morn": 276.27
          },
          "pressure": 1020,
          "humidity": 81,
          "dew_point": 276.77,
          "wind_speed": 3.06,
          "wind_deg": 294,
          "weather": [
            {
              "id": 500,
              "main": "Rain",
              "description": "light rain",
              "icon": "10d"
            }
          ],
          "clouds": 56,
          "pop": 0.2,
          "rain": 0.62,
          "uvi": 1.93
        },
      ]
    })
    const wrapper = shallow(<WeatherWidget weatherRecords={weatherData} />);

    expect(wrapper.html()).not.toBe(null);
  });
})