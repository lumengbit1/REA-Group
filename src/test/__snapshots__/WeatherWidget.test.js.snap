// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render snapshop testing 1.renders correctly 1`] = `ShallowWrapper {}`;

exports[`render snapshop testing 4.weatherRecords props has data 1`] = `ShallowWrapper {}`;

exports[`render snapshop testing renders correctly 1`] = `ShallowWrapper {}`;

exports[`render snapshop testing renders correctly 2`] = `ShallowWrapper {}`;
